pkg/customresourcestate/generate/generator/testdata/foo_types.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2023 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Changes to this file may require to regenerate the `foo-config.yaml`. Otherwise the
+// tests in ../generate_integration_test.go may fail.
+// The below marker can be used to regenerate the `foo-config.yaml` file by running
+// the following command:
+// $ go generate ./pkg/customresourcestate/generate/generator/testdata
+//go:generate sh -c "go run ../../../../../ generate ./... > foo-config.yaml"
+
+// +groupName=bar.example.com
+package foo
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// FooSpec is the spec of Foo.
+type FooSpec struct {
+	// This tests that defaulted fields are stripped for v1beta1,
+	// but not for v1
+	DefaultedString string `json:"defaultedString"`
+}
+
+// FooStatus is the status of Foo.
```

Comment:
Is it possible to unify `+Metrics:gauge`, `+Metrics: stateset` and `+Metrics:info:name` into one?

A metric has these:
1. metric name
2. metric value
3. key-value labels.

Reply:
xref back to: https://github.com/kubernetes/kube-state-metrics/issues/1899#issuecomment-1375445300

I think we should not merge them. This helps to distinct the three types we have in kube-state-metric's custom resource configuration:

https://github.com/kubernetes/kube-state-metrics/blob/main/pkg/customresourcestate/config_metrics_types.go#L24-L26

Having gauge and info being magically the same and having stateset separate seems not to be a clear UX for me and resolves to some magic happening in the back.

Reply:
I think the existing UX has a learning curve to know the differences between gauge/info/stateset.

This model is easier to understand:
```
1. metric name
2. metric value
3. key-value labels
```

We have two options:
1. Merge this PR and migrate to the above model in the future
2. Use the new model API now, while it's implemented by gauge/info/stateset

cc @dgrisonnet to share opinions on it

Reply:
To summarise a bit:

## Current implementation on this PR

The current model comes down to having three markers:
* `+Metrics:gauge`
* `+Metrics:info`
* `+Metrics:stateset`

Each of them maps to one type of the [configuration](https://github.com/kubernetes/kube-state-metrics/blob/main/pkg/customresourcestate/config_metrics_types.go#L24-L26) and by that also to the structs in the configuration file. 
By mapping the markers structs to the configuration file structs, the implementation of the markers gets simple and easy to maintain.
The generator aspect itself is still pretty complex though, although we are using controller-gen packages.

The help output currently is the following:
```sh
go run ./ generate -w foo

Metrics

+Metrics:gauge type
        Defines a Gauge metric and uses the implicit path to the field joined by the provided JSONPath as path for the metric configuration.
        [JSONPath=<string>]
        [help=<string>]
        [labelFromKey=<string>]
        [labelsFromPath=<>]
        name=<string>
        [nilIsZero=<bool>]
        [valueFrom=<string>]
...
+Metrics:info type
        Defines a Info metric and uses the implicit path to the field as path for the metric configuration.
        [JSONPath=<string>]
        [help=<string>]
        [labelFromKey=<string>]
        [labelsFromPath=<>]
        name=<string>
...
+Metrics:stateset type
        Defines a StateSet metric and uses the implicit path to the field as path for the metric configuration.
        [JSONPath=<string>]
        [help=<string>]
        [labelName=<string>]
        [labelsFromPath=<>]
        list=<[]string>
        name=<string>
...
```

From the configuration file perspective, these three types got introduced via #1777 to solve the following issues:
* #1755
* #1762

## Other proposed approach

If I get your proposal right: The other approach would be to have a single marker which is able to match all use cases.

Let's sum the above: variables/arguments we have / want:

* for every type:
  * `JSONPath`
  * `help`
  * `labelsFromPath`
  * `name`
* Only for `gauge` and `info`: 
  * `labelsFromPath`
* Only for `gauge`:
  * `nilIsZero`
* Only for `stateset`:
  * `list`

The help definition for the marker would look like this to fulfil all use-cases (+/- naming):

```
+Metrics:metric type
        Defines a metric ....
        [JSONPath=<string>]
        [help=<string>]
        [labelFromKey=<string>]
        [labelsFromPath=<>]
        [list=<[]string>]
        name=<string>
        [nilIsZero=<bool>]
        [valueFrom=<string>]
```

I don't think that this will be easier to understand from a UX perspective than the above.
Also from implementation perspective this will get more complicated:

* we would need logic to decide wether we have a gauge, info or stateset metric
  * identifying a stateset would be pretty easy: if `list` is defined it is a stateset (but what if it is defined by "accident" / copy/paste issue?)
  * determining if we want to have a gauge or info could work by checking if JSONPath points to a struct or array
    * if so: this is a info metric, if no: this is a gauge
    * **but:** technically this is very hard to implement, at least controller-gen does not really help us here. We would have to implement this logic from scratch AFAIK. This also means more and complex code to maintain.
* we would have to implement validation logic to check that we don't get arguments which exclude each other:
  * e.g. `list` and `nilIsZero` is defined
* The auto-generated help will not be obvious anymore when considering the variants. It would be a single entry listing all arguments.

## Appendix / personal opinion

I agree, at the end all comes down to:

> 1. metric name
> 2. metric value
> 3. key-value labels

But the additional aspects I see is:

* Programatically setting a metric value:
  * stateset use-case: "1 only if it matches a given string from a list, 0 otherwise"
  * gauge use-case: "if value is nil, use 0, no metric otherwise"
  * info use-case: "always 1"
* Maintainability / simplicity of the code

From learning perspective I still think it is good to split the different types, which come from the [OpenMetrics types definition](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#metric-types).
Documentation can build on top of the metric types and go through a "what are the arguments and what do they do" for each type.
For someone not being deep into kube-state-metrics custom resources, but having background from OpenMetrics it may be more obvious to get to the right parts of the documentation then.

After writing this: I think your question maybe is more about evolving the configuration file instead.

Reply:
PS: sorry for that long comment 😂  I hope this helps for the discussion. I am totally open to approaches, aspects, etc. I might miss.

I also agree 100% on this:

> I think the existing UX has a learning curve to know the differences between gauge/info/stateset.
> 
> This model is easier to understand:
> 
> ```
> 1. metric name
> 2. metric value
> 3. key-value labels
> ```

But I currently just don't see how we could build it easier without dropping features.